### PR TITLE
Fix the retrieval of plugins

### PR DESCRIPTION
**Pull request by pyrco**

### CODE REVIEW

**Comment by pyrco** [comment]
I think that is a bit too different scope for this change. It also needs a bit more thinking and designing. So for another time.

File: dissect/target/plugin.py [hunk 1]
```diff
@@ -540,74 +531,160 @@ def decorator(obj):
     return decorator
 
 
-def plugins(osfilter: Optional[type[OSPlugin]] = None) -> Iterator[PluginDescriptor]:
-    """Retrieve all plugin descriptors.
+def plugins(
```
Comment (cecinestpasunepipe) [inline dissect/target/plugin.py:534]: `target-query ~/TestSystemen/SCHARDT.img -n -q -f os.windows._os`

has now different behaviour (compared to main).

**Review by cecinestpasunepipe — CHANGES_REQUESTED** [review]

File: dissect/target/plugin.py [hunk 2]
```diff
@@ -540,74 +531,160 @@ def decorator(obj):
...
-                    continue
+    An exception to these exceptions is in the case of ``OSPlugin`` (sub-)class
+    plugins and ``os_filter`` is not ``None``. These plugins live in the
+    ``PLUGINS`` tree under the ``_os`` special key. Those plugins are only
+    returned if they fully match the provided ``osfilter``.
 
-                yield obj
+    The ``only_special_keys`` option returns only the plugins which are under a
+    special key that is defined in ``special_keys``. All filtering here will
+    happen as stated in the bove cases.
```
Comment (cecinestpasunepipe) [inline dissect/target/plugin.py:560]:
```suggestion
    happen as stated in the above cases.
```

File: dissect/target/plugin.py [hunk 3]
```diff
@@ -540,74 +531,160 @@ def decorator(obj):
...
 
-def _special_plugins(special_key: str) -> Iterator[PluginDescriptor]:
-    """Retrieve plugin descriptors stored under ``special_key``."""
+    if osfilter is not None:
+        # The PLUGINS tree does not include the hierarchy up to the plugins
+        # directory (dissect.target.plugins) for the built-in plugins. For the
+        # plugins in the directory specified in --plugin-path, the hierarchy
+        # starts at that directory.
+        #
+        # Build-in OSPlugins do have the dissect.target.plugins path in their
```
Comment (cecinestpasunepipe) [inline dissect/target/plugin.py:577]:
```suggestion
        # Built-in OSPlugins do have the dissect.target.plugins path in their
```

https://www.oxfordlearnersdictionaries.com/definition/english/built-in

File: dissect/target/plugin.py [hunk 4]
```diff
@@ -540,74 +531,160 @@ def decorator(obj):
...
+            # only traversed into if the module_path matches in full the
+            # (beginning of) the filter_path.
+            if (
+                not filter_path
+                or filter_path
+                and (
+                    issubclass(osfilter, general.default.DefaultPlugin)
+                    or "_os" not in obj
+                    or ("_os" in obj and (module_path == filter_path or module_path == filter_path[: len(module_path)]))
+                )
```
Comment (cecinestpasunepipe) [inline dissect/target/plugin.py:628]:
```suggestion

            osfilter_is_a_default_plugin =  issubclass(osfilter, general.default.DefaultPlugin)
            no_os_in_obj =  "_os" not in obj
            os_full_match_begin = "_os" in obj and (module_path == filter_path or module_path == filter_path[: len(module_path)])

            if (not filter_path or filter_path and (osfilter_is_a_default_plugin or no_os_in_obj or os_full_match_begin)):
```

Given the complexity, maybe use intermediate variables?

File: dissect/target/plugin.py [hunk 5]
```diff
@@ -540,74 +531,160 @@ def decorator(obj):
...
-                yield from _walk(obj)
+            module_path = prev_module_path.copy()
+            module_path.append(key)
+
+            # A branch in the PLUGINS tree is traveresed to the next level if
+            # there are no filters (which in effect means all plugins are
+            # returned).
+            # If there is a filter, a branch is traversed if there is no `_os`
+            # item on the next level. If there there is an `_os` item, it is
+            # only traversed into if the module_path matches in full the
```
Comment (cecinestpasunepipe) [inline dissect/target/plugin.py:619]:
```suggestion
            # only traversed into if the module_path matches in full, the
```
Comma reads easier.

**Review by cecinestpasunepipe — CHANGES_REQUESTED** [review]

File: dissect/target/plugin.py (hunk 4 above)
Comment (pyrco) [reply dissect/target/plugin.py:628]: I made the statement less complex in a different way.

**Review by cecinestpasunepipe — APPROVED** [review]

**Pull request merged**